src/pkg/artifact/dao/dao.go

```diff
@@ -282,6 +284,53 @@ func (d *dao) DeleteReferences(ctx context.Context, parentID int64) error {
 	return err
 }
 
+func (d *dao) ListWithLatest(ctx context.Context, query *q.Query) (artifacts []*Artifact, err error) {
+	ormer, err := orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sql := `SELECT a.*
+	FROM artifact a
+	JOIN (
+		SELECT repository_name, MAX(push_time) AS latest_push_time
+	FROM artifact
+	WHERE project_id = ? and %s = ?
+	GROUP BY repository_name
+	) latest ON a.repository_name = latest.repository_name AND a.push_time = latest.latest_push_time`
+
+	queryParam := make([]interface{}, 0)
+	var ok bool
+	var pid interface{}
+	if pid, ok = query.Keywords["ProjectID"]; !ok {
+		return nil, errors.New(nil).WithCode(errors.BadRequestCode).
+			WithMessage(`the value of "ProjectID" must be set`)
+	}
+	queryParam = append(queryParam, pid)
+
+	var attributionValue interface{}
+	if attributionValue, ok = query.Keywords["media_type"]; ok {
+		sql = fmt.Sprintf(sql, "media_type")
```

Comment:
Since media_type and artifact_type are user input parameters and are passed directly into the SQL here, is there a potential risk of sql injection?

Reply:
The SQL contains media_type and and artifact_type, but they are fixed strings here, not from the user input, no risk of SQL injection.

Reply:
as @stonezdj mentioned, there is no risk.